teps/0013-graceful-sidecar-termination.md

```diff
@@ -0,0 +1,221 @@
+
+Currently, the sidecar feature in a tekton pipeline/taskrun is not behaving like a typical kubernetes sidecar should. 
+The sidecar gets terminated immediately when the pipeline fails or completes without giving the sidecar enough time to 
+do what it originally intended to do (for example, upload logs, process artifacts etc.)
+In some cases the sidecar should be able to choose when to terminate itself. A sidecar is just a kubernetes container. 
+It's able to terminate itself with the command completion.
+
+
+This proposal is introducing a flag `waitForSidecar` which is a per-sidecar configuration that will allow users to set
+the nature of termination of a sidecar inside a task.
+
+## Motivation
+
+The first motivation is the counter intuitive nature of the sidecars. As a developer, I did not expect my sidecar to be
+terminated in such way. I had to check the docs to find out that the sidecar was being terminated using a Nop Image.
+
+Having no way to alter the behavior was the second motivation.
+
+Consider this task:
+Task:
+	- Step 1: Git Clone
+	- Step 2: Helm Diff/Configure
+	- Step 3: Helm Apply/Deploy
+	- Step 4: E2E
+		- Generates E2E artifacts
+		- Exposes pod and container metadata using downwards API
+
+	- Sidecar 1
+
+		The author of the task adds `Sidecar 1` that does the following:
+		- Polls the status of `Step 4` using downwards API.
+		- When the `Step 4` container is terminated, regardless of the reason of termination (failed when e2e fails and completed when e2e passes), it process all the artifacts generated by `Step 4` (Generates HTML report, does trend building etc.) and uploads the resultant artifacts to a permanent storage (S3).
+
+However, due to the current design, it terminates the sidecar immediately in case the E2E step fails. I.e., the `Step 4` container is terminated without success. 
+
+### Goals
+
+The goal of this proposal is to provide users more control over their sidecar container termination and make Tekton more flexible for complex implementation
+scenarios without introducing any breaking changes.
+
+The Nop Image termination can still be considered a good feature since in most cases, the sidecar does not need to run more than the pipelines lifetime and
+a non optimized injected sidecar may hang the pipeline indefinitely.
+
+### Non-Goals
+This proposal is not introducing any breaking change to current behavior. Sidecars still terminate using Nop Image by default when the pipeline is completed/failed.
+
+## Proposal
+
+The proposal is that we’ll make it a “per-sidecar” configuration on whether Tekton should determine and force terminate the sidecars once task run is done. I.e.; all tekton steps in the task are completed. It will default to `false` so that the new mechanism will only take effect when explicitly set by the flag in Sidecar configuration.
+
+
+### User Stories
+
+A `Pipeline` author can add a Sidecar with `waitForSidecar` enabled (true) and expect `natural death` of his/her sidecar.
```

Comment:
This is a design detail I think.  User Stories might be something like:

> As a Task author I want to add a sidecar that uploads the results of tests to external storage after the tests complete so that I can review the results later.

> As a Task author I want to add a sidecar that builds a report from the outcomes of some Steps and submits that report when the Task has reached a completed state so that the reports can be aggregated by an external system.

Reply:
Love it!